addons/website_slides/controllers/main.py

```diff
@@ -643,18 +684,107 @@ def _slide_channel_prepare_review_values(self, channel):
 
         return values
 
-    @http.route('/slides/channel/enroll', type='http', auth='public', website=True)
-    def slide_channel_join_http(self, channel_id):
-        # TDE FIXME: why 2 routes ?
+    @http.route('/slides/<int:channel_id>/invite', type='http', auth='public', website=True, sitemap=False)
+    def slide_channel_invite(self, channel_id, partner_id=False, invite_hash=False):
+        """ This route is included in the invitation link in email to join the course. It is the main entry point
+        for sharing course to attendees or to invite them directly.
+
+        It acts as a dipatcher, here are the supported use cases:
+            - If the current user (public or logged) can access the course, redirects to the course page
+            - If no user is logged and the public user has no access to the course, redirects to login page,
+            then redirects in turn to this route to be dispatched again, once the user is logged, according
+            to its access rights.
+            - If a logged user has not access to the course, redirects to the main page with all courses, with an
+            invite_state in the url that will trigger an error message to explain the user what is the problem.
+
+        :param channel_id: The id of the course the user is invited to. Do not use <model> in the route
+                           instead, otherwise an error 403 could be returned if the (public) user has no
+                           access to the record.
+
+        :return: redirects to the relevant route, see above.
+        """
+
+        def default_routing(channel=False, has_rights=False, invite_state=''):
+            """ This sub-method redirects to either the course or the main slides page with an appropriate error keyword."""
+            if not channel or not has_rights:
+                return request.redirect("/slides?invite_state=%s" % (invite_state or ('no_channel' if not channel else 'no_rights')))
+            return request.redirect("/slides/%s" % (slug(channel)))
+
+        # ----- Parameters
+        partner_id = int(partner_id) if partner_id else False
+        channel = request.env['slide.channel'].browse(int(channel_id)).exists()
+        if not channel:
+            return default_routing()
+
+        # ----- Compute rights of current user
+        has_rights = True
+        try:
+            channel.check_access_rights('read')
+            channel.check_access_rule('read')
+        except AccessError:
+            has_rights = False
+
+        # ----- Non-Specific Invitation or Non published channel
+        if not partner_id or not invite_hash:
+            # Generic invite
+            return default_routing(channel, has_rights)
```

Comment:
Works since we would have returned earlier without channel.
```suggestion
            return redirect_slides_main('no_rights')
```